panic: SIGSEGV while creating container

### Nomad version
Nomad v0.6.0-dev (56cd8b69132ece337834826d94cfe29a0c4f3119+CHANGES), provided by @dadgar in response to #2668

### Operating system and Environment details
CentOS 7.2
Docker 1.13.1
Consul 0.8.1

### Issue
The Nomad client fails with SIGSEGV while trying to create a container under some circumstances. This appears to be caused by the client making a request to list containers while a container is in the process of being removed. When the request for the specific container is made, it no longer exists, and the error handling (ironically) causes a panic, because the container var is nil inside the condition:

https://github.com/hashicorp/nomad/blob/56cd8b69132ece337834826d94cfe29a0c4f3119/client/driver/docker.go#L1098-L1103
https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L1196-L1201

```go
// Inspect the container and if the container isn't dead then return
// the container
container, err := client.InspectContainer(container.ID)
if err != nil {
	return nil, recoverableErrTimeouts(fmt.Errorf("Failed to inspect container %s: %s", container.ID, err))
}
```

### Reproduction steps
Unknown... this happens rarely

### Nomad Client logs (if appropriate)

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xba0a1b]
goroutine 5219617 [running]:
github.com/hashicorp/nomad/client/driver.(*DockerDriver).createContainer(0xc42efce700, 0xc424270450, 0x2e, 0xc42dbdd6c0, 0xc43731dc00, 0x0, 0x0, 0x0, 0xc42dbdd6c0, 0xc43731dc00, ...)
	/opt/gopath/src/github.com/hashicorp/nomad/client/driver/docker.go:1102 +0x9eb
github.com/hashicorp/nomad/client/driver.(*DockerDriver).Start(0xc42efce700, 0xc426f44580, 0xc423959180, 0xc432c5c5e0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/driver/docker.go:527 +0x691
github.com/hashicorp/nomad/client.(*TaskRunner).startTask(0xc42036b500, 0xc424acd860, 0x0)
	/opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:1366 +0x50a
github.com/hashicorp/nomad/client.(*TaskRunner).run(0xc42036b500)
	/opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:1061 +0x334
github.com/hashicorp/nomad/client.(*TaskRunner).Run(0xc42036b500)
	/opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:581 +0x556
created by github.com/hashicorp/nomad/client.(*AllocRunner).Run
	/opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:678 +0x9d9
```

--- a/client/driver/docker.go
+++ b/client/driver/docker.go
@@ -1179,10 +1179,10 @@ CREATE:
 		// container names with a / pre-pended to the Nomad generated container names
 		containerName := "/" + config.Name
 		d.logger.Printf("[DEBUG] driver.docker: searching for container name %q to purge", containerName)
-		for _, container := range containers {
+		for _, shimContainer := range containers {
 			d.logger.Printf("[DEBUG] driver.docker: listed container %+v", container)
 			found := false
-			for _, name := range container.Names {
+			for _, name := range shimContainer.Names {
 				if name == containerName {
 					found = true
 					break
@@ -1195,9 +1195,15 @@ CREATE:
 
 			// Inspect the container and if the container isn't dead then return
 			// the container
-			container, err := client.InspectContainer(container.ID)
+			container, err := client.InspectContainer(shimContainer.ID)
 			if err != nil {
-				return nil, recoverableErrTimeouts(fmt.Errorf("Failed to inspect container %s: %s", container.ID, err))
+				err = fmt.Errorf("Failed to inspect container %s: %s", shimContainer.ID, err)
+
+				// This error is always recoverable as it could
+				// be caused by races between listing
+				// containers and this container being removed.
+				// See #2802
+				return nil, structs.NewRecoverableError(err, true)
 			}
 			if container != nil && (container.State.Running || container.State.FinishedAt.IsZero()) {
 				return container, nil

